### PR TITLE
Revert previous attempt at Triton patch; use CustomCacheManger approach instead.

### PR DESCRIPTION
**Pull request by tdoublep**
I tested the previous fix for the Triton cache collision issue (see: #34) and it didn't work. 

I now see errors like:
```
FileNotFoundError: [Errno 2] No such file or directory: '/home/vllm/.triton/cache/1feb415f3280ca46eea8c4407a58c23e/fused_moe_kernel.json.tmp.pid_72_c0a0033e-6147-4520-ae3a-3847d02598f8'
```
which now shows the `uuid` instead of a random integer, but problem remains. 

This PR implements a different workaround, proposed by @cyang49, that tells Triton to use a custom cache manager which assigns a different directory based on the process id. 

This time I have tested it and it seems to work.

### CODE REVIEW

File: triton_patch/custom_cache_manager.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+            self.lock_path = os.path.join(self.cache_dir, "lock")
+            os.makedirs(self.cache_dir, exist_ok=True)
+        elif override:
+            self.cache_dir = default_override_dir()
+            self.cache_dir = os.path.join(self.cache_dir, self.key)
+        else:
+            # create cache directory if it doesn't exist
+            self.cache_dir = os.getenv("TRITON_CACHE_DIR",
+                                       "").strip() or default_cache_dir()
+            self.cache_dir = self.cache_dir + "_" + str(os.getpid())
```
Comment (njhill) [inline triton_patch/custom_cache_manager.py:24]: nit:
```suggestion
            self.cache_dir = f"{self.cache_dir}_{os.getpid()}"
```

File: triton_patch/custom_cache_manager.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+            print(f"{self.cache_dir=}")
```
Comment (njhill) [inline triton_patch/custom_cache_manager.py:25]: Delete this now? (or could make it a more descriptive message)

**Review by njhill — COMMENTED** [review]
Thanks @tdoublep!

Would be good for @tjohnson31415 to review too

**Comment by tjohnson31415** [comment]
I did a little investigation into this issue; this is what I found:

The stacktrace pointed to the error as coming from CompiledKernel.__init__() where the code lists the files and then tries to read each of them ([REF](
https://github.com/triton-lang/triton/blob/958fccea74da58e7e0595ab88ae6cd3f6795a173/python/triton/compiler/compiler.py#L229-L231)). Those files are created as artifacts of the compilation process in the code [here](
https://github.com/triton-lang/triton/blob/958fccea74da58e7e0595ab88ae6cd3f6795a173/python/triton/runtime/cache.py#L115-L127).

What is interesting is that the error includes `.tmp.pid` in the name which is a temporary file that the compilation is written to before doing an `os.replace`. I think the issue is that, between listing the files and trying to read the files, the `.tmp.pid` file is `oc.replace`d by another process; hence leading to a `FileNotFoundError`. Even with unique filenames for each process, multiple processes are using the same cache directory.

So a fix in the Triton code could be to remove the temporary files from the listing or to have the temp files created in a separate directory. That way, the compiled artifacts can still be shared across processes (though with some redundant compilations).

That said, I think this fix with a custom cache manager giving each process its own directory will also work and is a good work-around until a fix can be merged into Triton.

**Pull request merged**